core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1239 @@
+        y0: 0.54,
+        xLabel: 0.3875,
+        yLabel: 0.6,
+        id: 'State5',
+        label: 'State5',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State4: {
+        depth: 1,
+        offset: 3,
+        reachable: true,
+        x0: 0.7475,
+        y0: 0.33999999999999997,
+        xLabel: 0.8375,
+        yLabel: 0.4,
+        id: 'State4',
+        label: 'State4',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State3: {
+        depth: 1,
+        offset: 2,
+        reachable: true,
+        x0: 0.5225000000000001,
+        y0: 0.33999999999999997,
+        xLabel: 0.6125,
+        yLabel: 0.4,
+        id: 'State3',
+        label: 'State3',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State2: {
+        depth: 1,
+        offset: 1,
+        reachable: true,
+        x0: 0.29750000000000004,
+        y0: 0.33999999999999997,
+        xLabel: 0.3875,
+        yLabel: 0.4,
+        id: 'State2',
+        label: 'State2',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      State1: {
+        depth: 1,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.33999999999999997,
+        xLabel: 0.1625,
+        yLabel: 0.4,
+        id: 'State1',
+        label: 'State1',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      Introduction: {
+        depth: 0,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.14,
+        xLabel: 0.1625,
+        yLabel: 0.2,
+        id: 'Introduction',
+        label: 'Introduction',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      },
+      End: {
+        depth: 3,
+        offset: 0,
+        reachable: true,
+        x0: 0.07250000000000001,
+        y0: 0.7400000000000001,
+        xLabel: 0.1625,
+        yLabel: 0.8,
+        id: 'End',
+        label: 'End',
+        height: 0.12,
+        width: 0.18000000000000002,
+        reachableFromEnd: false
+      }
+    };
+    let graphWidthUpperBound = sgls.getGraphWidth(
+      AppConstants.MAX_NODES_PER_ROW, AppConstants.MAX_NODE_LABEL_LENGTH);
+    let graphHeight = sgls.getGraphHeight(nodeData);
+
+    // Here, nodeData has position values (x0, xLabel, width etc.) in terms of
+    // fractions of total height and width. modifyPositionValues converts these
+    // values to pixels.
+    let nodeDataWithPositionValueInPixel = sgls.modifyPositionValues(
+      nodeData, graphWidthUpperBound, graphHeight);
+    let expectedGraphBoundaries = {
+      bottom: 245.8,
+      left: 40.675000000000004,
+      right: 589.325,
+      top: 34.2
+    };
+
+    let expectedWidth = expectedGraphBoundaries.left +
```

Comment:
Thanks for the [explanation](https://github.com/oppia/oppia/pull/12946#discussion_r644413438) on this. There are a few things that confuse me about it:

- The explanation refers to a specific case where one node is at the center. But the test doesn't mention this special case at all. If you shift the node to the left then both leftEdge and rightEdge would decrease and so leftEdge + rightEdge < W; the opposite happens when you shift the node to the right. So I think I am confused about whether this statement you're making is supposed to hold in general.
- There should be some explanation about why the node data used above would "naturally" be likely to result in a graph that exceeds the upper bound. Otherwise you might just have a weak test case here. One way you can demonstrate it (in addition to adding comments) is to check the width for graphs of sizes N - 1 and N (where N is at the boundary of the max number of nodes in a row) and showing that these increase, then checking the width for a similar graph of size N + 1 and demonstrating that it is the same as the one for N.

Reply:
- I get your point. And I think the width should rather be `rightEdge` **-** `leftEdge`.
- Also, the function `graphWidth()` does not returns the actual width of the graph but the upper bound (i.e. 630). So, we would need to calculate the width using the above relation (`rightEdge` **-** `leftEdge`). But then, this width is never equal to 630, even when there are N nodes ( max number of nodes in a row possible), this difference is due to padding fraction. So, we won't be able to demonstrate the N-1, N, N+1 example.

So, the check should be that the width and height must always be less than 630 and graphHeight(), respectively. Does this sound okay?

Reply:
But how would you check that? It's impossible to check an "always" criterion with finite tests.

But if, as in my previous suggestion, you show that there's a steadily increasing sequence (N-1, N) that then plateaus out at a constant value, that seems like a reasonable way to check it. I didn't say that the width needs to be equal to 630.

Reply:
Added the checks